data_model/src/events/data/filters.rs

```diff
@@ -0,0 +1,616 @@
+//! This module contains `EventFilter` and entities for filter
+
+use super::*;
+
+/// Create entity filter type. See usage below
+///
+/// This could be implemented with generic struct, but it's bad to have nested generics in filters
+/// public API (especially for schemas)
+macro_rules! entity_filter {
+    (pub struct $name:ident { event: $entity_type:ty, filter: $event_filter_type:ty, }) => {
+        entity_filter! {
+            $name,
+            $entity_type,
+            $event_filter_type,
+            concat!("Filter for ", stringify!($event_filter_type), " entity"),
+            concat!("Construct new ", stringify!($name))
+        }
+    };
+    ($name:ident, $entity_type:ty, $event_filter_type:ty, $struct_doc:expr, $new_doc:expr) => {
+        #[derive(
+            Clone,
+            PartialOrd,
+            Ord,
+            PartialEq,
+            Eq,
+            Debug,
+            Decode,
+            Encode,
+            Deserialize,
+            Serialize,
+            IntoSchema,
+            Hash,
+        )]
+        #[doc = $struct_doc]
+        pub struct $name {
+            id_filter: FilterOpt<IdFilter<<$entity_type as Identifiable>::Id>>,
+            event_filter: FilterOpt<$event_filter_type>,
+        }
+
+        impl $name {
+            #[doc = $new_doc]
+            pub fn new(
+                id_filter: FilterOpt<IdFilter<<$entity_type as Identifiable>::Id>>,
+                event_filter: FilterOpt<$event_filter_type>,
+            ) -> Self {
+                Self {
+                    id_filter,
+                    event_filter,
+                }
+            }
+        }
+
+        impl Filter for $name {
+            type Item = $entity_type;
+
+            fn filter(&self, entity: &Self::Item) -> bool {
+                self.id_filter.filter(entity.id()) && self.event_filter.filter(entity)
+            }
+        }
+    };
+}
+
+#[cfg(feature = "roles")]
+mod role {
+    //! This module contains filters related to `RoleEvent`
+
+    use super::*;
+
+    entity_filter!(
+        pub struct RoleFilter {
+            event: RoleEvent,
+            filter: RoleEventFilter,
+        }
+    );
```

Comment:
```suggestion
    entity_filter!(RoleFilter, RoleEvent, RoleEventFilter);
```
This could be API for macro. You prefer the current API?

Reply:
I'm trying to follow this [rule](https://rust-lang.github.io/api-guidelines/macros.html):

> Rust macros let you dream up practically whatever input syntax you want. Aim to keep input syntax familiar and cohesive with the rest of your users' code by mirroring existing Rust syntax where possible. Pay attention to the choice and placement of keywords and punctuation.

Reply:
I'll be completely honest, I think it should be a proc macro with derive annotations. I prefer keeping the `struct`, because often LSP isn't good enough at finding definitions through macro expansion…

Reply:
Also relevant: 
> Macros that produce a struct or enum as output should support attributes so that the output can be used with derive.

```
bitflags! {
    #[derive(Default, Serialize)]
    struct Flags: u8 {
        const ControlCenter = 0b001;
        const Terminal = 0b010;
    }
}
```

Reply:
Iroha is meant to be edited just as much as linked against.